Framework/API/src/ExperimentInfo.cpp

```diff
@@ -414,44 +428,63 @@ void ExperimentInfo::swapInstrumentParameters(Geometry::ParameterMap &pmap) {
  */
 void ExperimentInfo::cacheDetectorGroupings(const det2group_map &mapping) {
   populateIfNotLoaded();
-  m_detgroups.clear();
+  if (mapping.empty()) {
+    cacheDefaultDetectorGrouping();
+    return;
+  }
+  setNumberOfDetectorGroups(mapping.size());
+  size_t specIndex = 0;
   for (const auto &item : mapping) {
-    m_det2group[item.first] = m_detgroups.size();
-    m_detgroups.push_back(item.second);
+    m_det2group[item.first] = specIndex;
+    setDetectorGrouping(specIndex, item.second);
+    specIndex++;
   }
-  // Create default grouping if `mapping` is empty.
-  cacheDefaultDetectorGrouping();
 }
 
-/// Returns the detector IDs that make up the group that this ID is part of
-const std::set<detid_t> &
-ExperimentInfo::getGroupMembers(const detid_t detID) const {
+/** Sets the number of detector groups.
+ *
+ * This method should not need to be called explicitly. The number of detector
+ * groups will be set either when initializing a MatrixWorkspace, or by calling
+ * `cacheDetectorGroupings` for an ExperimentInfo stored in an MDWorkspace. */
+void ExperimentInfo::setNumberOfDetectorGroups(const size_t count) const {
   populateIfNotLoaded();
-  auto iter = m_det2group.find(detID);
-  if (iter != m_det2group.end()) {
-    return m_detgroups[iter->second];
-  } else {
-    throw std::runtime_error(
-        "ExperimentInfo::getGroupMembers - Unable to find ID " +
-        std::to_string(detID) + " in lookup");
-  }
+  if (m_spectrumInfo)
+    m_spectrumDefinitionNeedsUpdate.clear();
```

Comment:
Maybe a similar thing to the `invalidateSpectrumDefinition` required here. Not necessary, but you might consider going further the `std::vector<char>` could be made into a separate class with these methods attached.